### PR TITLE
[C] do not generate code without main reactor

### LINKED ISSUE #913: Files without main don't compile if they have imports
The file `test/C/lib/Imported.lf` gives compile errors when compiled with lfc or Epoch. The reason is that it has an `import` statement that is ignored. This should be easy to fix (?).

### PR DESCRIPTION
**Pull request by housengw**
Do not generate code when there is no main reactor

Fixes #913

### CODE REVIEW

**Comment by lhstrh** [comment]
How does this PR relate to #913? Not generating code when there is no main reactor means that we cannot generate any pure library code. Is that really what we want?

**Comment by housengw** [comment]
The error outlined in #913 is due to `CGenerator` not checking if main reactor exist before generating code in `doGenerate`.

The `Cpp` target also doesnt generate code when there is no main reactor. It might make sense to standardize code generation behavior across targets.

**Comment by cmnrd** [comment]
I think we should really generate a library if there is no main reactor. In fact, I think we should create a library for all LF files and only generate a small binary that uses this library if there is a main reactor. Note that for C++ we have a long open issue: #83.

For the short term solution, though, it might be better to just forbid compilation without a main reactor. Without a package system, it does not make much sense to compile LF programs as libraries.

File: org.lflang/src/org/lflang/generator/GeneratorUtils.java [hunk 1]
```diff
@@ -55,16 +56,14 @@ private GeneratorUtils() {
...
-            if (targetDecl != null) {
-                throw new InvalidSourceException("There is more than one target!"); // FIXME: check this in validator
-            }
             targetDecl = t;
+            cnt++;
         }
-        if (targetDecl == null) {
-            throw new InvalidSourceException("No target found!");
-        }
+        if (cnt > 1)  throw new InvalidSourceException("There is more than one target!"); // FIXME: check this in validator
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/GeneratorUtils.java:65]:
```suggestion
```
I don't think our grammar permits more than one target declaration, so this check is unnecessary altogether.

File: org.lflang/src/org/lflang/generator/GeneratorUtils.java [hunk 2]
```diff
@@ -55,16 +56,14 @@ private GeneratorUtils() {
      * Return the target declaration found in the given resource.
      */
     public static TargetDecl findTarget(Resource resource) {
+        int cnt = 0;
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/GeneratorUtils.java:59]:
```suggestion
```

File: org.lflang/src/org/lflang/generator/GeneratorUtils.java [hunk 3]
```diff
@@ -55,16 +56,14 @@ private GeneratorUtils() {
...
         TargetDecl targetDecl = null;
         for (TargetDecl t : findAll(resource, TargetDecl.class)) { // getAllContents should never return null.
-            if (targetDecl != null) {
-                throw new InvalidSourceException("There is more than one target!"); // FIXME: check this in validator
-            }
             targetDecl = t;
+            cnt++;
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/GeneratorUtils.java:63]:
```suggestion
```

File: org.lflang/src/org/lflang/generator/GeneratorUtils.java [hunk 4]
```diff
@@ -55,16 +56,14 @@ private GeneratorUtils() {
      * Return the target declaration found in the given resource.
      */
     public static TargetDecl findTarget(Resource resource) {
+        int cnt = 0;
         TargetDecl targetDecl = null;
         for (TargetDecl t : findAll(resource, TargetDecl.class)) { // getAllContents should never return null.
-            if (targetDecl != null) {
-                throw new InvalidSourceException("There is more than one target!"); // FIXME: check this in validator
-            }
             targetDecl = t;
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/GeneratorUtils.java:62]:
```suggestion
            return t;
```

File: org.lflang/src/org/lflang/generator/GeneratorUtils.java [hunk 5]
```diff
@@ -55,16 +56,14 @@ private GeneratorUtils() {
...
+            cnt++;
         }
-        if (targetDecl == null) {
-            throw new InvalidSourceException("No target found!");
-        }
+        if (cnt > 1)  throw new InvalidSourceException("There is more than one target!"); // FIXME: check this in validator
+        if (cnt == 0) throw new InvalidSourceException("No target found!");
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/GeneratorUtils.java:66]:
```suggestion
```
Our grammar _requires_ a target declaration.

**Review by lhstrh — APPROVED** [review]
Looks like a good quick fix. Like @cmnrd said, more work would be needed for the generation of library code. I've made suggestions for the removal of checks that I think are unnecessary.

**Comment by lhstrh** [comment]
Oh, and congrats on filing PR #1000 :partying_face:

File: org.lflang/src/org/lflang/generator/GeneratorUtils.java [hunk 6]
```diff
@@ -56,14 +56,10 @@ private GeneratorUtils() {
      * Return the target declaration found in the given resource.
      */
     public static TargetDecl findTarget(Resource resource) {
-        int cnt = 0;
         TargetDecl targetDecl = null;
         for (TargetDecl t : findAll(resource, TargetDecl.class)) { // getAllContents should never return null.
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/GeneratorUtils.java:60]:
```suggestion
```

File: org.lflang/src/org/lflang/generator/GeneratorUtils.java [hunk 7]
```diff
@@ -56,14 +56,10 @@ private GeneratorUtils() {
...
-            targetDecl = t;
-            cnt++;
+            return t;
         }
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/GeneratorUtils.java:62]:
```suggestion
```

File: org.lflang/src/org/lflang/generator/GeneratorUtils.java [hunk 8]
```diff
@@ -56,14 +56,10 @@ private GeneratorUtils() {
      * Return the target declaration found in the given resource.
      */
     public static TargetDecl findTarget(Resource resource) {
-        int cnt = 0;
         TargetDecl targetDecl = null;
         for (TargetDecl t : findAll(resource, TargetDecl.class)) { // getAllContents should never return null.
-            targetDecl = t;
-            cnt++;
+            return t;
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/GeneratorUtils.java:61]: Wouldn't this also work?
```suggestion
            return findAll(resource, TargetDecl.class).iterator().next();
```

**Review by lhstrh — APPROVED** [review]
One more suggestion.

File: org.lflang/src/org/lflang/generator/GeneratorUtils.java (hunk 8 above)
Comment (housengw) [reply org.lflang/src/org/lflang/generator/GeneratorUtils.java:61]: this would also work.

**Review by lhstrh — APPROVED** [review]

**Pull request merged**